nRF5340 external-xip DFU sample (#13509)
nrfconnect/sdk-nrf: samples/nrf5340/smp_svr/README.rst

Author: This sample is basing on upstream sample which we can refer to.
I do believe that it would be possible that we will refer upstream Read.me here and describe things which are specific for external-xip point of view.

.. _smp_svr_ext_xip:

SMP Server Sample with external XIP
###################################

Overview
********

This sample implements a Simple Management Protocol (SMP) server that demonstrates splitting the application that it partially resides on internal flash and partially on QSPI flash.

SMP is a basic transfer encoding for use with the MCUmgr management protocol.
For more information about MCUmgr and SMP, please see :ref:`device_mgmt`.

Refer to zephyr-rtos :zephyr:code-sample:`smp-svr` for information about SMP features implemented by the sample.

> Thes samples enables the spliting the application into intenal and external part thasnks to:

Reviewer: Thes->These ?
"These samples demonstrate how an application can be split into internal and external part, where the internal part is executed from SoC internal storage and the external part is executed, with aid of the XIP feature a QSPI peripheral, from external memory"?

> * Enabled by :kconfig:option:`CONFIG_XIP_SPLIT_IMAGE` Split XIP image feature, which turns on the build system level support for image division and adapts image signing support for MCUboot.

Reviewer: "The split XIP image feature is enabled by :kconfig:option:`CONFIG_XIP_SPLIT_IMAGE`; the Kconfig option enables the build ..."

> * Project CMake which describes relocation of part of libraries or object to the external XIP area.

Reviewer: "to an external XIP", it is some area probably.

> * Linker scripts file which describes external QSPI area to which code should be linked.

Reviewer: "script file describing" I think sounds better than "which describes".
I think what the script does is it defines SoC RAM range reserved for mapping of external memory device, connected to the QSPI peripheral of SoC, which in turn allows you to build a code that is relocated to that mapped range.

* Static partition manager description for appropriate memory mapping.

Building the sample application
*******************************

The below steps describe how to build and run the ``smp_svr`` sample in

> NCS.

Reviewer: No point to break the line.

To build the sample:

   .. code-block:: console

      west build \
         -b nrf5340dk_nrf5340_cpuapp \
         nrf/samples/nrf5340/smp_svr

> NCS build system building all exexutables required by the sample: the application divided in part for internal flash and external QSPI flash part, the networking stack, the MCUboot.

Reviewer: "The NCS build system will build all the required binaries for the platform: "
", the MCUboot" -> "and the MCUboot".

> Then it signing MCUboot's images: the internal applicatrion part image, the external application part image, the networking image.

Reviewer: "The build process includes signing all of the application binaries except from MCUboot that does not require signing in this configuration."

Flashing the sample image
*************************

To upload the bundle of images and the MCUboot, we simply use ``west flash``

> like normal.

Reviewer: No need for line break, probably no need for simply.

For flashing the sample:

.. code-block:: console

    west flash

Sample image: hello world!
==========================

> The ``smp_svr`` app is ready to run.  Just reset your board and test the app

Reviewer: Double " ".

behaviour", and it performs upload to the "image-1", the MCUboot secondary slot.

For clarity, here is PM label to slot to ``<image>`` translation table:

    +-----------------------+--------+-------------+---------------------------+
    | PM label              | Slot   | -n <image>  |         covers            |
    +=======================+========+=============+===========================+
    | "mcuboot_secondary"   | slot-1 |     0       | internal application part |
    +-----------------------+--------+-------------+---------------------------+
    | "mcuboot_secondary_1" | slot-3 |     1       | networking                |
    +-----------------------+--------+-------------+---------------------------+
    | "mcuboot_secondary_2" | slot-5 |     2       | qspi application part     |
    +-----------------------+--------+-------------+---------------------------+

.. note::

   The ``-e`` option actually means "no erase", and is provided to the mcumgr
   to prevent it from sending erase command to target, before updating image.
   The options is always needed when ``-n`` is used for image selection,

>    as the erase command is hardcoded to erase slot-1 ("image-1"),

Reviewer: It is hardcoded to erase "secondary slot", which in case of DirectXIP execution will be the non-running/opposite slot.

Author: This is not DirectXIP mode.

   regardless of which slot is uploaded at the time.

Upload the signed image
=======================

To upload the signed image, use the following command:

.. code-block:: console

   sudo mcumgr <connection string> image upload build/zephyr/<signed_upload.bin> -e -n <image>

   use:
   `internal_flash_update.bin` file and <image> = 0. for updating the internal application part image.
   `net_core_app_update.bin` file and <image> = 1. for updating the networking core image
   `qspi_flash_update.bin` file and <image> = 2. for updating the qspi application part image.

.. note::

>    <image> 0 and 2 are respective the application internal flash part image and

Reviewer: "<image> values 0 and 2 correspond to the internally and the externally linked parts of application, respectively."

   the application external QSPI flash part image. These two images need to be
   uploaded always as pairs as they are linked together.

Test the image
==============

In order to instruct MCUboot to swap the images we need to test the image first,
obtain image SHA:

.. code-block:: console

   imgtool verify build/zephyr/<signed_upload.bin>

then make sure it boots using the image digest fetched:

.. code-block:: console

   sudo mcumgr <connection string> image test <hash of image>

Need set to test all images which should be update.
After that MCUBoot will swap these images on the next reset.

.. note::

>    The sample supports only MCUBoot overwrite-only mode - so ther is no diference whether given image is marked to be update by either `test` or `confirm`.

Reviewer: I think that "overwrite-only" should be explained here.

Reset remotely
==============

We can reset the device remotely to observe (use the console output) how
MCUboot swaps the images:

.. code-block:: console

   sudo mcumgr <connection string> reset

> Upon reset MCUboot will write to primary slots the secondary slots images.

Reviewer: I think this line can be somehow merged with 161-162.

   regardless of which slot is uploaded at the time.

Upload the signed image
=======================

To upload the signed image, use the following command:

.. code-block:: console

   sudo mcumgr <connection string> image upload build/zephyr/<signed_upload.bin> -e -n <image>

   use:
   `internal_flash_update.bin` file and <image> = 0. for updating the internal application part image.
   `net_core_app_update.bin` file and <image> = 1. for updating the networking core image
   `qspi_flash_update.bin` file and <image> = 2. for updating the qspi application part image.

.. note::

   <image> values 0 and 2 correspond to the internally and the externally linked parts of application, respectively.
   These two images need to be uploaded always in pairs as they are linked together.

Test the image
==============

In order to instruct MCUboot to swap the images we need to test the image first,
obtain image SHA:

.. code-block:: console

   imgtool verify build/zephyr/<signed_upload.bin>

then make sure it boots using the image digest fetched:

.. code-block:: console

   sudo mcumgr <connection string> image test <hash of image>

> The selected mode, in this case test, needs to be applied to all images that constitute the complete application.

Reviewer 2: ```suggestion
The selected mode in this test case must be applied to all images that constitute the complete application.
```

After that MCUBoot will swap these images on the next reset.

.. note::

   The sample supports only MCUBoot overwrite-only mode. In that mode the old image is overwritten by the content of the incoming image during firmware update process.

>    Therefore ther is no diference whether given image is marked to be update by either `test` or `confirm`.

Reviewer 2: ```suggestion
   Therefore, there is no difference if the given image is marked to be update by either ``test`` or ``confirm``.
```

Reset remotely

> ==============

Reviewer 2: If this is part of "Testing" section, please set to 1 subheading lower:
```suggestion
-------------
```

We can reset the device remotely and observe (use the console output) how
MCUboot upgarding the images:

.. code-block:: console

>    sudo mcumgr <connection string> reset

Reviewer 2: Please see [NCS sample template](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/sample_README.html#references), and check if a "References" or "Dependencies" section is needed. Otherwise, disregard this comment.

Author: I've added "Dependencies".

To upload the signed image, use the following command:

.. code-block:: console

   sudo mcumgr <connection string> image upload build/zephyr/<signed_upload.bin> -e -n <image>

   use:
   `internal_flash_update.bin` file and <image> = 0. for updating the internal application part image.
   `net_core_app_update.bin` file and <image> = 1. for updating the networking core image
   `qspi_flash_update.bin` file and <image> = 2. for updating the qspi application part image.

.. note::

   <image> 0 and 2 are respective the application internal flash part image and
   the application external QSPI flash part image. These two images need to be
   uploaded always as pairs as they are linked together.

Test the image
==============

In order to instruct MCUboot to swap the images we need to test the image first,
obtain image SHA:

.. code-block:: console

   imgtool verify build/zephyr/<signed_upload.bin>

then make sure it boots using the image digest fetched:

.. code-block:: console

   sudo mcumgr <connection string> image test <hash of image>

Need set to test all images which should be update.
After that MCUBoot will swap these images on the next reset.

.. note::

>    Currently mcumgr client doesn't support `image list` for 3 images configuration.

Author: info on networking primary slot is unavailable as its area_id direct to nothing (application doesn't have access to networking flash).

Reset remotely
==============

We can reset the device remotely to observe (use the console output) how
MCUboot swaps the images:

.. code-block:: console

   sudo mcumgr <connection string> reset

Upon reset MCUboot will swap primary and secondary slots.

Confirm new image
=================

The new images are now loaded into primary slots, but they will be swapped back into secondary

> on the next reset unless these image are confirmed. To confirm the new image:

Author: My mistake - that is overwrite only mode.

.. _smp_svr_sample:

SMP Server Sample with external XIP
###################################

Overview
********

> This sample application implements a Simple Management Protocol (SMP) server whitch demonstrating spliting the appliucationso that it partially resides on internal flash and partially on QSPI flash.

Reviewer 3: ```suggestion
This sample implements a Simple Management Protocol (SMP) server that demonstrates splitting the application that it partially resides on internal flash and partially on QSPI flash.
```

> The appication uses original zephyr-rtos smp_svr sample code.

Reviewer 3: ```suggestion
```
Irrelevant IMHO.

> Refer to zephyr-rtos sample documnetation for information about SMP features implemented by the sample.

Reviewer 3: Add link to upstream documentation.

Author: @annwoj How to add link to sdk-zephyr doc?

Reviewer 4: Please try using: `:zephyr:code-sample:`smp-svr`, in the current state I think it's just missing the first colon

The mcumgr ``image upload`` command may be provided optional ``-e -n <image>`` parameter that will
select target image for upload; when parameter is no provided, 0 is assumed, which means "default
behaviour", and it performs upload to the "image-1", the MCUboot secondary slot.

The Partition manager (PM) label for slot-to-``<image>`` translation is shown in the table below:

    +-----------------------+--------+-------------+---------------------------+
    | PM label              | Slot   | -n <image>  |       Firmware part       |
    +=======================+========+=============+===========================+
    | "mcuboot_secondary"   | slot-1 |     0       | internal application part |
    +-----------------------+--------+-------------+---------------------------+
    | "mcuboot_secondary_1" | slot-3 |     1       | networking                |
    +-----------------------+--------+-------------+---------------------------+
    | "mcuboot_secondary_2" | slot-5 |     2       | qspi application part     |
    +-----------------------+--------+-------------+---------------------------+

.. note::

>    The ``-e`` option means "no erase", and is provided to the Mcumgr to prevent it from sending the erase command to target before updating image.

Reviewer 4: ```suggestion
   The ``-e`` option means "no erase", and is provided to the MCUmgr to prevent it from sending the erase command to the target before updating the image.
```

   This options is always needed when ``-n`` is used for image selection, as the erase command is hardcoded to erase slot-1 ("image-1"), regardless of which slot is uploaded at the time.

Upload the signed image
-----------------------

> To upload the signed image, use the following command:

Reviewer 4: ```suggestion
To upload the signed image, run the following command:
```

.. code-block:: console

   sudo mcumgr <connection string> image upload build/zephyr/<signed_upload.bin> -e -n <image>

> Use:

Reviewer 4: ```suggestion
Adjust the command depending on the case:
```

>    `internal_flash_update.bin` file and <image> = 0. for updating the internal application part image.
>    `net_core_app_update.bin` file and <image> = 1. for updating the networking core image
>    `qspi_flash_update.bin` file and <image> = 2. for updating the qspi application part image.

Reviewer 4: ```suggestion
* Use the :file:`internal_flash_update.bin` file and ``<image> = 0`` to update the internal application image.
* Use the :file:`net_core_app_update.bin` file and ``<image> = 1`` to update the networking core image.
* Use :file:`qspi_flash_update.bin` file and ``<image> = 2`` for updating the QSPI application image.
```
@nvlsianpu edited this a bit, please double-check if the code brackets are applied correctly

> 

Reviewer 4: ```suggestion
```

.. note::

>    The <image> values ``0`` and ``2`` correspond to the internally and the externally linked parts of application, respectively.

Reviewer 4: ```suggestion
   The ``<image>`` values correspond to the internally (``0``) and externally (``2``) linked parts of the application.
```
@nvlsianpu, @kittydepa was trying something here to make it more straightforward but might've done the opposite - do you think this edit is clear enough? If not then I have no strong feeling when it comes to keeping it

>    These two images must always be uploaded in together, becuse they are linked together.

Reviewer 4: ```suggestion
   These two images are linked, therefore, must always be uploaded together.
```

Test the image
==============

> To instruct MCUboot to swap the images, you must test the image first by obtaining the image's SHA with the following command:

Reviewer 4: ```suggestion
To initiate the image swap with MCUboot, first you must test the image by obtaining its SHA with the following command:
```

.. code-block:: console

   imgtool verify build/zephyr/<signed_upload.bin>

> Next ensure it boots using the image digest fetched by running followin command:

Reviewer 4: ```suggestion
Ensure it boots using the image digest fetched by running the following command:
```

.. code-block:: console

   sudo mcumgr <connection string> image test <hash of image>

The selected mode in this test case must be applied to all images that constitute the complete application.
After that, MCUBoot will swap these images during the next reset.

.. note::

>    The sample supports only supports MCUBoot overwrite-only mode.
>    In that mode the old image is overwritten by the content of the incoming image during firmware update process.
>    Therefore there is no difference if the given image is marked to be update by either ``test`` or ``confirm``.

Reviewer 4: ```suggestion
   The sample supports the MCUBoot overwrite-only mode.
   In that mode, the old image is overwritten by the content of the incoming image during the firmware update process.
   Therefore, if an image is marked for update as either ``test`` or ``confirm``, the process is not affected.
```
Edited the last part a bit - meaning tbc

Reset remotely
--------------

> You can reset the device remotely and observe the console output to see how MCUboot has upgraded the images by running the following command:

Reviewer 4: ```suggestion
To check how MCUboot has upgraded the images, you can reset the device remotely by running the following command:
```
